File: ios/Sources/Account/Services/AppReviewService.swift
import StoreKit
import SwiftUI
import os.log

private let logger = Logger(subsystem: "com.listenai", category: "AppReview")

/// Service for managing App Store review prompts.
/// Uses a feedback flow: first asks if user is enjoying, then triggers native review if yes.
@MainActor
final class AppReviewService: ObservableObject {
    static let shared = AppReviewService()

    // MARK: - Keys for UserDefaults

    private enum Keys {
        static let lastReviewPromptDate = "appReview_lastPromptDate"
        static let reviewPromptCount = "appReview_promptCount"
        static let hasCompletedReview = "appReview_hasCompleted"
        static let hasCompletedSuccessfulJourney = "appReview_hasCompletedJourney"
    }

    // MARK: - Configuration

    /// Minimum days between review prompts
    private let minimumDaysBetweenPrompts = 30

    /// Maximum review prompts to show in app lifetime
    private let maxPrompts = 3

    // MARK: - Notifications

    /// Posted when the feedback prompt should be shown
    static let shouldShowFeedbackPromptNotification = Notification.Name("AppReviewService.shouldShowFeedbackPrompt")

    // MARK: - State

    @Published var showFeedbackPrompt = false

    private init() {}

    /// Check and trigger feedback prompt if appropriate (posts notification if should show)
    func checkAndTriggerFeedbackPrompt() {
        if shouldShowFeedbackPrompt() {
            recordFeedbackPromptShown()
            NotificationCenter.default.post(name: Self.shouldShowFeedbackPromptNotification, object: nil)
            logger.info("Posted feedback prompt notification")
        }
    }

    // MARK: - Public API

    /// Call this after a successful voice clone creation
    func recordVoiceCloneSuccess() {
        UserDefaults.standard.set(true, forKey: Keys.hasCompletedSuccessfulJourney)
        logger.info("Voice clone success recorded - user journey complete")
    }

    /// Call this after first successful TTS playback completion
    func recordTTSPlaybackSuccess() {
        UserDefaults.standard.set(true, forKey: Keys.hasCompletedSuccessfulJourney)
        logger.info("TTS playback success recorded - user journey complete")
    }

    /// Check if we should show the feedback prompt
    func shouldShowFeedbackPrompt() -> Bool {
        // Don't show if user has already completed a review
        if UserDefaults.standard.bool(forKey: Keys.hasCompletedReview) {
            logger.debug("Skipping feedback: user has completed review")
            return false
        }

        // Don't show if we've reached max prompts
        let promptCount = UserDefaults.standard.integer(forKey: Keys.reviewPromptCount)
        if promptCount >= maxPrompts {
            logger.debug("Skipping feedback: max prompts reached (\(promptCount))")
            return false
        }

        // Check if enough time has passed since last prompt
        if let lastPromptDate = UserDefaults.standard.object(forKey: Keys.lastReviewPromptDate) as? Date {
            let daysSinceLastPrompt = Calendar.current.dateComponents([.day], from: lastPromptDate, to: Date()).day ?? 0
            if daysSinceLastPrompt < minimumDaysBetweenPrompts {
                logger.debug("Skipping feedback: only \(daysSinceLastPrompt) days since last prompt")
                return false
            }
        }

        // Check if user has completed at least one successful journey (voice clone OR TTS playback)
        if !UserDefaults.standard.bool(forKey: Keys.hasCompletedSuccessfulJourney) {
            logger.debug("Skipping feedback: no successful journey completed yet")
            return false
        }

        logger.info("Should show feedback prompt")
        return true
    }

    /// Record that we showed a feedback prompt
    func recordFeedbackPromptShown() {
        let promptCount = UserDefaults.standard.integer(forKey: Keys.reviewPromptCount)
        UserDefaults.standard.set(promptCount + 1, forKey: Keys.reviewPromptCount)
        UserDefaults.standard.set(Date(), forKey: Keys.lastReviewPromptDate)
        logger.info("Feedback prompt shown. Total prompts: \(promptCount + 1)")
    }

    /// User responded positively - trigger native App Store review
    func requestAppStoreReview() {
        logger.info("User enjoying app - requesting App Store review")

        // Mark as completed so we don't ask again
        UserDefaults.standard.set(true, forKey: Keys.hasCompletedReview)

        // Request the native App Store review
        if let scene = UIApplication.shared.connectedScenes.first(where: { $0.activationState == .foregroundActive }) as? UIWindowScene {
            SKStoreReviewController.requestReview(in: scene)
        }
    }

    /// User responded negatively - could open feedback form in future
    func userNotEnjoying() {
        logger.info("User not enjoying - skipping review request")
        // Just record we asked and don't prompt again for a while
        // In future, could open a feedback form here
    }
}

// MARK: - Feedback Prompt View

/// Simple "Are you enjoying?" prompt that appears before triggering App Store review
struct FeedbackPromptView: View {
    let onYes: () -> Void
    let onNo: () -> Void

    @Environment(\.dismiss) private var dismiss

    var body: some View {
        VStack(spacing: 24) {
            // Icon
            Image(systemName: "heart.fill")
                .font(.system(size: 48))
                .foregroundStyle(.pink)

            // Title
            Text("Enjoying ReadAloud?")
                .font(.title2.bold())
                .foregroundStyle(.primary)

            // Subtitle
            Text("Your feedback helps us improve the app for everyone.")
                .font(.body)
                .foregroundStyle(.secondary)
                .multilineTextAlignment(.center)
                .padding(.horizontal, 24)

            // Buttons
            HStack(spacing: 16) {
                // No button
                Button {
                    onNo()
                    dismiss()
                } label: {
                    Text("Not Really")
                        .font(.headline)
                        .foregroundStyle(.primary)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 14)
                        .background(Color(.systemGray5))
                        .clipShape(RoundedRectangle(cornerRadius: 12))
                }

                // Yes button
                Button {
                    onYes()
                    dismiss()
                } label: {
                    Text("Yes!")
                        .font(.headline)
                        .foregroundStyle(.white)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 14)
                        .background(Color.black)
                        .clipShape(RoundedRectangle(cornerRadius: 12))
                }
            }
            .padding(.horizontal, 24)
        }
        .padding(.vertical, 32)
        .background(Color(.systemBackground))
    }
}

#Preview {
    FeedbackPromptView(
        onYes: { print("Yes tapped") },
        onNo: { print("No tapped") }
    )
}

File: android/app/src/main/java/com/listenai/service/review/AppReviewService.kt
package com.listenai.service.review

import android.app.Activity
import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.google.android.play.core.review.ReviewManagerFactory
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Date
import java.util.concurrent.TimeUnit

/**
 * Service for managing Google Play Store review prompts.
 * Uses a feedback flow: first asks if user is enjoying, then triggers native review if yes.
 *
 * Mirrors iOS AppReviewService behavior:
 * - Rate limits prompts (max 3 prompts, minimum 30 days between)
 * - Triggers after voice clone success or TTS playback start
 * - Shows custom feedback prompt before native review
 */
class AppReviewService private constructor(private val context: Context) {

    companion object {
        private const val TAG = "AppReviewService"

        // SharedPreferences keys
        private const val PREFS_NAME = "app_review_prefs"
        private const val KEY_LAST_PROMPT_DATE = "appReview_lastPromptDate"
        private const val KEY_PROMPT_COUNT = "appReview_promptCount"
        private const val KEY_HAS_COMPLETED_REVIEW = "appReview_hasCompleted"
        private const val KEY_HAS_COMPLETED_JOURNEY = "appReview_hasCompletedJourney"

        // Configuration
        private const val MINIMUM_DAYS_BETWEEN_PROMPTS = 30
        private const val MAX_PROMPTS = 3

        @Volatile
        private var INSTANCE: AppReviewService? = null

        fun getInstance(context: Context): AppReviewService {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: AppReviewService(context.applicationContext).also { INSTANCE = it }
            }
        }
    }

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    // State flow for UI observation
    private val _shouldShowFeedbackPrompt = MutableStateFlow(false)
    val shouldShowFeedbackPrompt: StateFlow<Boolean> = _shouldShowFeedbackPrompt.asStateFlow()

    /**
     * Check and trigger feedback prompt if appropriate.
     * Emits to shouldShowFeedbackPrompt state flow if conditions are met.
     */
    fun checkAndTriggerFeedbackPrompt() {
        if (shouldShowFeedbackPromptInternal()) {
            recordFeedbackPromptShown()
            _shouldShowFeedbackPrompt.value = true
            Log.i(TAG, "Triggering feedback prompt")
        }
    }

    /**
     * Dismiss the feedback prompt (called after user responds or dismisses)
     */
    fun dismissFeedbackPrompt() {
        _shouldShowFeedbackPrompt.value = false
    }

    // MARK: - Public API

    /**
     * Call this after a successful voice clone creation
     */
    fun recordVoiceCloneSuccess() {
        prefs.edit().putBoolean(KEY_HAS_COMPLETED_JOURNEY, true).apply()
        Log.i(TAG, "Voice clone success recorded - user journey complete")
    }

    /**
     * Call this after first successful TTS playback start
     */
    fun recordTTSPlaybackSuccess() {
        prefs.edit().putBoolean(KEY_HAS_COMPLETED_JOURNEY, true).apply()
        Log.i(TAG, "TTS playback success recorded - user journey complete")
    }

    /**
     * Check if we should show the feedback prompt (internal logic)
     */
    private fun shouldShowFeedbackPromptInternal(): Boolean {
        // Don't show if user has already completed a review
        if (prefs.getBoolean(KEY_HAS_COMPLETED_REVIEW, false)) {
            Log.d(TAG, "Skipping feedback: user has completed review")
            return false
        }

        // Don't show if we've reached max prompts
        val promptCount = prefs.getInt(KEY_PROMPT_COUNT, 0)
        if (promptCount >= MAX_PROMPTS) {
            Log.d(TAG, "Skipping feedback: max prompts reached ($promptCount)")
            return false
        }

        // Check if enough time has passed since last prompt
        val lastPromptDate = prefs.getLong(KEY_LAST_PROMPT_DATE, 0L)
        if (lastPromptDate > 0) {
            val daysSinceLastPrompt = TimeUnit.MILLISECONDS.toDays(Date().time - lastPromptDate)
            if (daysSinceLastPrompt < MINIMUM_DAYS_BETWEEN_PROMPTS) {
                Log.d(TAG, "Skipping feedback: only $daysSinceLastPrompt days since last prompt")
                return false
            }
        }

        // Check if user has completed at least one successful journey (voice clone OR TTS playback)
        if (!prefs.getBoolean(KEY_HAS_COMPLETED_JOURNEY, false)) {
            Log.d(TAG, "Skipping feedback: no successful journey completed yet")
            return false
        }

        Log.i(TAG, "Should show feedback prompt")
        return true
    }

    /**
     * Record that we showed a feedback prompt
     */
    private fun recordFeedbackPromptShown() {
        val promptCount = prefs.getInt(KEY_PROMPT_COUNT, 0)
        prefs.edit()
            .putInt(KEY_PROMPT_COUNT, promptCount + 1)
            .putLong(KEY_LAST_PROMPT_DATE, Date().time)
            .apply()
        Log.i(TAG, "Feedback prompt shown. Total prompts: ${promptCount + 1}")
    }

    /**
     * User responded positively - trigger native Google Play review
     */
    fun requestPlayStoreReview(activity: Activity) {
        Log.i(TAG, "User enjoying app - requesting Play Store review")

        // Mark as completed so we don't ask again
        prefs.edit().putBoolean(KEY_HAS_COMPLETED_REVIEW, true).apply()

        // Request the native Play Store review
        val reviewManager = ReviewManagerFactory.create(context)
        val request = reviewManager.requestReviewFlow()

        request.addOnCompleteListener { task ->
            if (task.isSuccessful) {
                val reviewInfo = task.result
                val flow = reviewManager.launchReviewFlow(activity, reviewInfo)
                flow.addOnCompleteListener {
                    Log.i(TAG, "Review flow completed")
                }
            } else {
                Log.e(TAG, "Failed to get review info: ${task.exception?.message}")
            }
        }
    }

    /**
     * User responded negatively - could open feedback form in future
     */
    fun userNotEnjoying() {
        Log.i(TAG, "User not enjoying - skipping review request")
        // Just dismiss the prompt and don't ask again for a while
        // In future, could open a feedback form here
    }
}
